### PR TITLE
Doip

### PR DESCRIPTION
**Pull request by wosym**

### CODE REVIEW

File: boards/nucleo-f207zg/include/periph_conf.h [hunk 1]
```diff
@@ -41,6 +41,7 @@ static const dma_conf_t dma_config[] = {
     { .stream = 4 },    /* DMA1 Stream 4 - SPI2_TX */
     { .stream = 14 },   /* DMA2 Stream 6 - USART6_TX */
     { .stream = 6 },    /* DMA1 Stream 6 - USART2_TX */
+    { .stream = 8 },    /* DMA2 Stream 0 - ETH_TX    */ // NOTE: Probeersel... komt uit andere periph_config!
```
Comment (toonst) [inline boards/nucleo-f207zg/include/periph_conf.h]: rebase on master to get latest periph_conf that includes this already

File: examples/doip/main.c [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    { NULL, NULL, NULL }
+};
+
+int main(void)
+{
+    int ret = 0;
+    char line_buf[SHELL_DEFAULT_BUFSIZE];
+
+    puts("Starting doip example");
+    xtimer_usleep(2000000);         //necesarry for some reason when sending UDP
```
Comment (toonst) [inline examples/doip/main.c]: there's `xtimer_sleep` which you can pass seconds to. Also use a preprocessor define for all magic values.

File: exp/Makefile [hunk 3]
```diff
@@ -0,0 +1,65 @@
+###
```
Comment (toonst) [inline exp/Makefile:1]: is this exp app still needed?

File: pkg/lwip/contrib/lwip.c [hunk 4]
```diff
@@ -42,11 +42,19 @@
...
+
 #include "lwip.h"
 
 #define ENABLE_DEBUG    (0)
 #include "debug.h"
 
+#ifdef USE_DHCP
+#include <lwip/dhcp.h>
+#endif
+
```
Comment (toonst) [inline pkg/lwip/contrib/lwip.c:57]: we should let the user decide whether to use DHCP or not.

File: examples/doip/main.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    doip_sa source = 0xe801;
+    doip_ta target = 0x1fff;
+    uint8_t uds_data[] = { UDS_SERVICES_TP, UDS_SUPRESS_REPLY };
+
+    while (1) {
+        doip_send_udp(source, target, DOIP_DIAGNOSTIC_MESSAGE, uds_data, 2,
+                      ip_addr);
+        doip_send_udp(0, 0, DOIP_VEHICLE_IDENTIFICATION_REQUEST, NULL, 0,
+                      ip_addr);
+        xtimer_usleep(1000000);
```
Comment (toonst) [inline examples/doip/main.c]: xtimer_sleep iso xtimer_usleep, use preprocessor defines for magic values.

File: examples/doip/main.c [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    if (argc < 3) {
+        msg = DOIP_VEHICLE_IDENTIFICATION_REQUEST;
+    }
+    if (msg == DOIP_VEHICLE_IDENTIFICATION_REQUEST) {
+        source = 0;
+        target = 0;
+        dlen = 0;
+    }
+    else {
+        source = 0xe801;
```
Comment (toonst) [inline examples/doip/main.c]: preprocessor defines for magic values

File: sys/Makefile [hunk 7]
```diff
@@ -160,6 +160,9 @@ endif
 ifneq (,$(filter netif,$(USEMODULE)))
     DIRS += net/netif
 endif
+ifneq (,$(filter doip,$(USEMODULE)))
+    DIRS += doip
+endif
```
Comment (toonst) [inline sys/Makefile:165]: I think doip belongs under `sys/net`instead of `sys`

File: sys/doip/Makefile [hunk 8]
```diff
@@ -0,0 +1,4 @@
+ifeq (gnu, $(TOOLCHAIN))
+  CFLAGS += -Wno-cast-function-type
```
Comment (toonst) [inline sys/doip/Makefile]: Please remove this hack

File: sys/doip/color.h [hunk 9]
```diff
@@ -0,0 +1,73 @@
+//Regular text
```
Comment (toonst) [inline sys/doip/color.h:1]: this belongs in the app, not in sys

File: sys/doip/doip.c [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+}
+
+int doip_send_udp(doip_sa sa, doip_ta ta, uint16_t payload_type, uint8_t *data,
+                  uint32_t dlen, char *ip_addr)
+{
+    int ret = 0;
+    sock_udp_ep_t remote = SOCK_IPV4_EP_ANY;
+    sock_udp_ep_t local = SOCK_IPV4_EP_ANY;
+    sock_udp_t sock;
```
Comment (toonst) [inline sys/doip/doip.c]: this sock can be made part of a `sock_doip_t`

File: sys/include/doip.h [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    DOIP_NO_SOCKET,
+    DOIP_ERROR
+} doip_result;
+
+
+
+/* ####################*/
+
+
+/* ####### DoIP Functions ########## */
```
Comment (toonst) [inline sys/include/doip.h:153]: I miss a `sock_doip_t`here, which holds the underlying udp socket, tcp socket, static buffers and other state. This socket can then be created by a `sock_doip_create` function and closed by a `sock_doip_close`function, similar to the other protocols which implement a socket interface. See eg. `sys/include/net/sock/udp.h`

The biggest benefit of this is that the socket contains all the state of a connection, allowing to open multiple connections.

File: tests/doip/main.c [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+    EMB_UNIT_TESTFIXTURES(fixtures) {
+        new_TestFixture(test_DoIP_Data_request__input_handling),
+    };
+    EMB_UNIT_TESTCALLER(doip_tests, NULL, NULL, fixtures);
+    return (Test *)&doip_tests;
+
+}
+
+int main(void)
```
Comment (toonst) [inline tests/doip/main.c:44]: This test app is definitely needed, so please fix this.

File: sys/include/doip.h [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+ * All DoIP layer services have the same general format. Service primitives are written in the form:
+ * service_name.type(parameter A, parameter B [, parameter C, ...])
+ *
+ * where “service_name” is the name of the service, e.g. _doip_data; “type” indicates the type of the service
+ * primitive; and “parameter A, parameter B, [parameter C, ...]” are the doip_SDU as a list of values passed by
+ * the service primitive. The brackets indicate that this part of the parameter list may be empty.
+ *
+ * Available service primitives:
+ * - request: request a service from the service provider
+ * - inidcation: service provider informs a service user about an internal event of the network layer or the service request of a peer protocol layer entity service user
```
Comment (toonst) [inline sys/include/doip.h]: typo: ìnidcation`

File: sys/doip/doip.c [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+        puts("Error creating UDP sock");
+        return 1;
+    }
+
+    if (data == NULL && dlen != 0) {
+        return -1;
+    }
+
+
+    //TODO: make a message_create function?
```
Comment (toonst) [inline sys/doip/doip.c]: yes! ;-)

File: sys/doip/doip.c [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+    local.port = 41123;
+
+    uint32_t payload_len = dlen + 4;        //TA and SA should also be included in the payload length!
+    //Not sure if it's always true though... What if TA and SA are 0 (e.g. with Veh. Id. Req)? Then those aren't added to payload, thus, no +4?
+
+    int msglen = 0;         //Total length of message to be sent over UDP
+
+    uint8_t buf[50] = { '\0' };
+    if (sock_udp_create(&sock, &local, NULL, 0)) {
```
Comment (toonst) [inline sys/doip/doip.c]: move this socket creation to a new `sock_doip_create`function

File: sys/doip/doip.c [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+            printf("0x%x ", buf[i]);
+        }
+        printf("\"\n");
+        doip_print_msg_parsed(buf, ret);
+        puts(reset);
+    }
+
+    puts("======================");
+    sock_udp_close(&sock);
+    xtimer_sleep(1);
```
Comment (toonst) [inline sys/doip/doip.c]: sleep still needed?

File: sys/doip/doip.c [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+        for (int i = 0; i < ret; i++) {
+            printf("0x%x ", buf[i]);
+        }
+        printf("\"\n");
+        doip_print_msg_parsed(buf, ret);
+        puts(reset);
+    }
+
+    puts("======================");
+    sock_udp_close(&sock);
```
Comment (toonst) [inline sys/doip/doip.c]: move closing of the socket to a new `socket_doip_close` function

File: sys/doip/doip.c (hunk 15 above)
Comment (toonst) [reply sys/doip/doip.c]: A doip socket can be used so the underlying udp socket doesn't have to be created and closed every time you want to send something

File: examples/doip/Makefile [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+CFLAGS += -DNETMASK=$(NETMASK)
+CFLAGS += -DGATEWAY=$(GATEWAY)
+USEMODULE += lwip_netdev
+USEMODULE += lwip_sock_udp
+USEMODULE += lwip lwip_sock_ip lwip_netdev
+USEMODULE += lwip_udp lwip_sock_udp
+USEMODULE += lwip_tcp lwip_sock_tcp
+USEMODULE += netdev_default
+USEMODULE += xtimer
+USEMODULE += shell
```
Comment (toonst) [inline examples/doip/Makefile:37]: Please group USEMODULE's and other stuff separately.

File: examples/doip/netutil.c [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+int start_dhcp(void)
+{
+    static netdev_t stm32eth;
+    extern void stm32_eth_netdev_setup(netdev_t *netdev);
+    static struct netif netif[1];
+
+    stm32_eth_netdev_setup(&stm32eth);
+    int ret = 0;
+
+    if (netif_add(&netif[0], IP4_ADDR_ANY, IP4_ADDR_ANY, IP4_ADDR_ANY,
```
Comment (toonst) [inline examples/doip/netutil.c]: netif add should be part of the lwip bootstrap code.

File: examples/doip/netutil.c [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+                  &stm32eth, lwip_netdev_init, tcpip_input) == NULL) {
+        puts("Could not add stm32_eth device\n");
+        return -1;
+    }
+    ret = dhcp_start(&netif[0]);
```
Comment (toonst) [inline examples/doip/netutil.c]: This should be done from the application. The netif can be retrieved with `netif_find()`

File: examples/doip/netutil.c [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+        for (int i = 0; i < LWIP_IPV6_NUM_ADDRESSES; i++) {
+            if (!ipv6_addr_is_unspecified((ipv6_addr_t *)&iface->ip6_addr[i])) {
+                printf(" inet6 %s\n",
+                       ipv6_addr_to_str(addrstr,
+                                        (ipv6_addr_t *)&iface->ip6_addr[i],
+                                        sizeof(addrstr)));
+            }
+        }
+#endif
+#ifndef APP_IPV6
```
Comment (toonst) [inline examples/doip/netutil.c]: Can we try to re-use the MODULE_LWIP_IPV6 instead of adding another preprocessor define?

**Review by toonst — CHANGES_REQUESTED** [review]
Finished review. I skipped the `exp` app because I think it is a leftover.

File: sys/include/doip.h (hunk 11 above)
Comment (wosym) [reply sys/include/doip.h:153]: @toonst Should the code be POSIX compliant? Because if I'm not mistaken, POSIX has _t reserved for their own structs.

Comment (toonst) [reply sys/include/doip.h:153]: it's fine to use _t here.

**Pull request closed without merging**